pkg/kubelet/certificate/certificate_manager.go

```diff
@@ -0,0 +1,371 @@
+	if m.certSigningRequestClient == nil {
+		glog.Infof("Kubernetes client is nil, not starting certificate manager.")
+		return
+	}
+
+	glog.Info("Starting to rotate expiring certificates.")
+	go wait.Forever(func() {
+		for range time.Tick(syncPeriod) {
+			err := m.rotateCerts()
+			if err != nil {
+				glog.Errorf("Could not rotate certificates: %v", err)
+			}
+		}
+	}, 0)
+}
+
+// cleanupPairs is responsible for recovering from any partial certificate
+// rotations rolling back to the most recent matching pair, and deleting any
+// keys that don't have a matching certificate.
+func (m *manager) cleanupPairs() error {
+	if err := m.removeUnmatchedPairs(); err != nil {
+		return err
+	}
+
+	keyPath := filepath.Join(m.keyDirectory, pairNamePrefix+keyExtension)
+	certPath := filepath.Join(m.certDirectory, pairNamePrefix+certExtension)
+	if fileExists(keyPath) && fileExists(certPath) {
+		// Make sure the selected key/cert pair are symlinks. This facilitates
+		// key rotation.
+		if err := fixSymlink(keyPath); err != nil {
+			return err
+		}
+		if err := fixSymlink(certPath); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// removeUnmatchedPairs will remove key files that don't have a matching cert
+// file, and cert files that don't have a matching key file.
+func (m *manager) removeUnmatchedPairs() error {
+	keyfiles, err := ioutil.ReadDir(m.keyDirectory)
+	if err != nil {
+		return err
+	}
+	keys := map[string]bool{}
+	for _, keyfile := range keyfiles {
+		if filepath.Ext(keyfile.Name()) == keyExtension {
+			keys[withoutExt(keyfile.Name())] = false
+		}
+	}
+	certfiles, err := ioutil.ReadDir(m.certDirectory)
+	if err != nil {
+		return err
+	}
+	for _, certfile := range certfiles {
+		if filepath.Ext(certfile.Name()) == certExtension {
+			if _, ok := keys[withoutExt(certfile.Name())]; ok {
+				keys[withoutExt(certfile.Name())] = true
+			} else {
+				certPath := filepath.Join(m.certDirectory, certfile.Name())
+				if err := os.Remove(certPath); err != nil {
+					glog.Errorf("The certificate %q doesn't have a matching key but can't be deleted: %v", certPath, err)
+				}
+			}
+		}
+	}
+	for filename, certFound := range keys {
+		if !certFound {
+			keyPath := filepath.Join(m.keyDirectory, filename+keyExtension)
+			if err := os.Remove(keyPath); err != nil {
+				glog.Errorf("The certificate %q doesn't have a matching key but can't be deleted: %v", keyPath, err)
+			}
+		}
+	}
+	return nil
+}
+
+func fileExists(file string) bool {
+	_, err := os.Stat(file)
+	return err == nil || os.IsExist(err)
+}
+
+// withoutExt returns the given filename after removing the extension. The
+// extension to remove will be the result of filepath.Ext().
+func withoutExt(filename string) string {
+	return strings.TrimSuffix(filename, filepath.Ext(filename))
+}
+
+// fixSymLink moves the given file to another location and replaces it with a
+// symlink to the new location.
+func fixSymlink(filename string) error {
```

Comment:
Changed this. The cert/key files are no longer moved or removed. They will be abandoned once rotation starts.